code/modules/power/groundmap_geothermal.dm

```diff
@@ -502,15 +434,92 @@ GLOBAL_VAR_INIT(active_bluespace_generators, 0)
 	for(var/obj/machinery/floor_warn_light/toggleable/generator/light AS in GLOB.generator_alarm_lights)
 		light.disable()
-	say("Bluespace restabilisation successful. Planetary assets protected.")
+
+/obj/machinery/power/geothermal/tbg/welder_act(mob/living/user, obj/item/I)
+	if(corrupted)
+		var/obj/item/tool/weldingtool/WT = I
+		if(user.skills.getRating(SKILL_ENGINEER) < SKILL_ENGINEER_ENGI)
+			user.visible_message(span_notice("[user] fumbles around figuring out the resin tendrils on [src]."),
+			span_notice("You fumble around trying to burn off the resin tendrils."))
+			user.balloon_alert(user, "You fumble around trying to burn off the resin tendrils.")
+			var/fumbling_time = 10 SECONDS - 2 SECONDS * user.skills.getRating(SKILL_ENGINEER)
+			if(!do_after(user, fumbling_time, NONE, src, BUSY_ICON_UNSKILLED, extra_checks = CALLBACK(WT, TYPE_PROC_REF(/obj/item/tool/weldingtool, isOn))))
+				return
+
+		if(!WT.remove_fuel(1, user))
+			to_chat(user, span_warning("You need more welding fuel to complete this task."))
+			return
+		playsound(loc, 'sound/items/weldingtool_weld.ogg', 25)
+		user.visible_message(span_notice("[user] carefully starts burning [src]'s resin off."),
+		span_notice("You start carefully burning the resin off."))
+		user.balloon_alert(user, "You start carefully burning the resin off.")
+		add_overlay(GLOB.welding_sparks)
+
+		if(!do_after(user, 20 SECONDS - clamp((user.skills.getRating(SKILL_ENGINEER) - SKILL_ENGINEER_ENGI) * 5, 0, 20) SECONDS, NONE, src, BUSY_ICON_BUILD, extra_checks = CALLBACK(WT, TYPE_PROC_REF(/obj/item/tool/weldingtool, isOn))))
+			cut_overlay(GLOB.welding_sparks)
+			return FALSE
+
+		playsound(loc, 'sound/items/welder2.ogg', 25, 1)
+		cut_overlay(GLOB.welding_sparks)
+		last_corrupted = corrupted
+		corrupted = 0
+		stop_processing()
+		update_icon()
+	return ..()
+
+/obj/machinery/power/geothermal/tbg/interact_hand(mob/living/user)
+	if(corrupted)
+		to_chat(user, span_warning("You have to clean that generator before it can be used!"))
+		return FALSE
+	return ..()
+
+/obj/machinery/power/geothermal/tbg/update_overlays()
+	. = ..()
+	if(corrupted)
+		. += image(icon, src, "overlay_corrupted", layer)
+
+/obj/machinery/power/geothermal/tbg/examine(mob/user, distance, infix, suffix)
+	. = ..()
+	if(corrupted)
+		. += "It is covered in writhing tendrils [!isxeno(user) ? "that could be cut away with a welder" : ""]."
+	if(!isxeno(user) && !is_corruptible)
+		. += "It is reinforced, making us not able to corrupt it."
+
+///Allow generators to generate psy points
+/obj/machinery/power/geothermal/tbg/proc/activate_corruption(datum/source)
+	SIGNAL_HANDLER
+	UnregisterSignal(SSdcs, list(COMSIG_GLOB_OPEN_TIMED_SHUTTERS_LATE, COMSIG_GLOB_OPEN_TIMED_SHUTTERS_XENO_HIVEMIND, COMSIG_GLOB_OPEN_SHUTTERS_EARLY, COMSIG_GLOB_TADPOLE_LANDED_OUT_LZ, COMSIG_GLOB_TADPOLE_RAPPEL_DEPLOYED_OUT_LZ))
+	corruption_on = TRUE
+	start_processing()
+
+/obj/machinery/power/geothermal/tbg/proc/corrupt(hivenumber)
```

Comment:
autodoc. Also new proc definitions go after existing ones, you have them all over the place throughout this file.